src/OrbitVulkanLayer/TimerQueryPool.h

```diff
@@ -76,30 +81,24 @@ class TimerQueryPool {
     return device_to_query_pool_.at(device);
   }
 
-  // Tries to find a free query slot in the device's pool. It returns `false` if no slot was found
-  // (all slots are occupied) and true otherwise.
-  // In case it successfully found a slot, the index will be written to the given `allocated_index`.
+  // Returns a free query slot from the device's pool if one still exists. It returns `false` if all
+  // slots are occupied and true otherwise. If successful, the index will be written to the given
+  // `allocated_index`.
   //
   // Note that the pool must be initialized using `InitializeTimerQueryPool` before.
   // See also `ResetQuerySlots` to make occupied slots available again.
   [[nodiscard]] bool NextReadyQuerySlot(VkDevice device, uint32_t* allocated_index) {
     absl::WriterMutexLock lock(&mutex_);
-    CHECK(device_to_potential_next_free_index_.contains(device));
     CHECK(device_to_query_slots_.contains(device));
-    uint32_t potential_next_free_slot = device_to_potential_next_free_index_.at(device);
-    std::vector<SlotState>& slots = device_to_query_slots_.at(device);
-    uint32_t current_slot = potential_next_free_slot;
-    do {
-      if (slots.at(current_slot) == SlotState::kReadyForQueryIssue) {
-        device_to_potential_next_free_index_[device] = (current_slot + 1) % num_timer_query_slots_;
-        slots.at(current_slot) = SlotState::kQueryPendingOnGpu;
-        *allocated_index = current_slot;
-        return true;
-      }
-      current_slot = (current_slot + 1) % num_timer_query_slots_;
-    } while (current_slot != potential_next_free_slot);
-
-    return false;
+    std::vector<uint32_t>& free_slots = device_to_free_slots_.at(device);
+    if (free_slots.empty()) {
+      return false;
+    }
+    *allocated_index = free_slots.back();
+    CHECK(device_to_query_slots_.at(device)[*allocated_index] == SlotState::kReadyForQueryIssue);
+    device_to_query_slots_.at(device)[*allocated_index] = SlotState::kQueryPendingOnGpu;
+    free_slots.pop_back();
```

Comment:
Nit/Personal preference: I would move the free_slots.pop_back up to the free_slots.back();

Reply:
Yes, agree. Done.

Reply:
Thanks!